meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+collectNeededInputsFromContract (Choice observation contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (When observation _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (While observation _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromObservation observation
+           , collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Scale value1 value2 value3 contract) = 
+  S.unions [ collectNeededInputsFromValue value1
+           , collectNeededInputsFromValue value2
+           , collectNeededInputsFromValue value3
+           , collectNeededInputsFromContract contract ]
+collectNeededInputsFromContract (Let _ contract1 contract2) =
+  S.unions [ collectNeededInputsFromContract contract1
+           , collectNeededInputsFromContract contract2 ]
+collectNeededInputsFromContract (Use _) = S.empty
+
+-- Add inputs and action ids to state.
+-- Return Nothing on redundant or irrelevant inputs
+addAnyInput :: BlockNumber -> AnyInput -> S.Set IdInput -> State -> Maybe State
+--  current block ---^
+addAnyInput _ (Action idInput) _ (State state)
+  | idInput `S.member` state.usedIds = Nothing
+  | true = Just (State (state { usedIds = S.insert idInput usedIdsSet }))
+  where usedIdsSet = state.usedIds
+addAnyInput _ (Input (IChoice idChoice choice)) neededInputs (State state)
+  | (WIdChoice idChoice) `M.member` state.choices = Nothing
+  | (IdChoice idChoice) `S.member` neededInputs = Just (State (state { choices = M.insert (WIdChoice idChoice) choice state.choices }))
+  | true = Nothing
+addAnyInput blockNumber (Input (IOracle idOracle timestamp value)) neededInputs (State state) =
+  case M.lookup idOracle oracleMap of
+    Just v ->
+        if (timestamp > v.blockNumber) && (timestamp <= blockNumber) && ((IdOracle idOracle) `S.member` neededInputs)
+        then Just newState
+        else Nothing
+    Nothing -> Just newState
+  where newState = State (state { oracles = M.insert idOracle {blockNumber: timestamp
+                                                              ,value: value} oracleMap })
+        oracleMap = state.oracles
+
+-- Decides whether something has expired
+isExpired :: BlockNumber -> BlockNumber -> Boolean
+isExpired currBlockNum expirationBlockNum = currBlockNum >= expirationBlockNum
+
+-- Expire commits
+expireOneCommit :: IdCommit -> CommitInfo -> CommitInfo
+expireOneCommit idCommit (CommitInfo commitInfo) =
+  CommitInfo
+  case M.lookup idCommit currentCommits of
+    Just v ->
+          let person = v.person in
+          let value = v.amount in
```

Comment:
Replace this with `Just { person, amount }` or `Just v@{ person, amount }`.

Reply:
Yes. I couldn't get the pattern to work. Will fix